drivers/SmartThings/sonos/src/utils.lua

```diff
@@ -0,0 +1,96 @@
+---@module 'utils'
+local utils = {}
+-- build a exponential backoff time value generator
+--
+-- max: the maximum wait interval (not including `rand factor`)
+-- inc: the rate at which to exponentially back off
+-- rand: a randomization range of (-rand, rand) to be added to each interval
+function utils.backoff_builder(max, inc, rand)
+  local count = 0
+  inc = inc or 1
+  return function()
+    local randval = 0
+    if rand then
+      randval = math.random() * rand * 2 - rand
+    end
+
+    local base = inc * (2 ^ count - 1)
+    count = count + 1
+
+    -- ensure base backoff (not including random factor) is less than max
+    if max then base = math.min(base, max) end
+
+    -- ensure total backoff is >= 0
+    return math.max(base + randval, 0)
+  end
+end
+
+function utils.labeled_socket_builder(label)
+  local log = require "log"
+  local socket = require "cosock.socket"
+  local ssl = require "cosock.ssl"
+
+  label = (label or "")
+  if #label > 0 then
+    label = label .. " "
+  end
+
+  local function make_socket(host, port, wrap_ssl)
```

Comment:
The only difference here between the Hue version of `make_socket` is that we don't set the TCP keepalive option. That's because Sonos only ever uses one-shot REST stuff to establish basic details and after that everything is websocket. So we don't want long-lived sockets.